libio/serial_com.cpp

```diff
@@ -182,18 +182,26 @@ size_t SerialCom::write(const uint8_t *buf, size_t len)
 	return res;
 }
 
-void SerialCom::flush()
+void SerialCom::flush(const std::chrono::milliseconds &timeout)
 {
 	unsigned char buf[256];
 	size_t len = sizeof(buf);
-	try{
+	auto now = std::chrono::steady_clock::now();
+	auto deadline = now + timeout;
+	try
+	{
 		// read remaining buffer
-		while (len == sizeof(buf))
+		while (len == sizeof(buf) && now < deadline)
+		{
 			len = read(buf, sizeof(buf));
+			now = std::chrono::steady_clock::now();
+		}
 	}
 	catch (const std::exception &e) {
 		std::cerr << e.what() << std::endl;
 	}
```

Comment:
Why did you decide to catch here and print them instead of propagating them further up the stack?
In this fashion, the caller wouldn't know about the error and the error isn't actually handled here, is it?

Reply:
I maybe have a bad vision of how to handle throw (I think this was my first try at it in this project). For me the low-level should catch the exception and decide what to do next, and high level need to catch only new more compact exceptions, that already combine several events having happened. Therefore this examples shows it
```
sc: read total of 256 bytes
sc: flushed 256 chars in 2000 ms
Device didn't stop streaming. Try to disconnect and reconnect.
sp: init failed:sp: failed to stop streaming
initialization failed
disconnecting
```
I would in fact throw in the flush a message like "device could not be flushed within timeout" because the flush should not deciding that it is a problem of stop streaming, as flush could be used for other purposes.
I catch the read throws, in the flush as it compromises the correct flushing. If the device got disconnected while flushing, it is not a flush error but a disconnect error.
Then the upper level decides what to do with it and says it failed to initialize due to a failed stop streaming (either due to flush timeout or disconnection)

Each of the messages (and throw) come from different level, on is `serial_com` the other `serial_protocol` and the last one is the node.

Reply:
> I would in fact throw in the flush a message like "device could not be flushed within timeout".

That's a much less helpful message for the user. If flushing fails that _always_ means that the device still continuously streams, which is obviously not expected when you call `flush()`.

> If the device got disconnected while flushing, it is not a flush error but a disconnect error.

So why then do you want to hide the disconnect error from the caller and pretend `flush()` finished w/o errors?

Reply:
I basically would like more message, maybe at different level, than a single throw leading to interpretation of the wrong reasons. if flush is called without stop streaming, but after a single triggered send (not streaming) and in the future, flush is used to clean spurious data after an invalid message, one would have to change the throw of the flush.
If flush did not finish due to disconnection, there is no reason to tell the error is a timeout in flush, if the timeout was not reached, the error that causes the driver to fail is the disconnection.

If you prefer to rework the throwing system, do, I will have to verify if I still understand what is happening when a device fails after the rework. Currently it was clear to me.

Reply:
> I basically would like more message, maybe at different level, than a single throw

So, you are asking actually for a stack trace at the highest level? There are tools for this, e.g. https://index.ros.org/p/backward_ros. No need to manually intercept and rethrow exceptions.

Currently, you are printing an error message in flush(), when receiving a read exception. But you don't report this as an error to the caller! This is what I am complaining about.

> If flush did not finish due to disconnection, there is no reason to tell the error is a timeout in flush

In that case there is no timeout?!?

> If the timeout was not reached, the error that causes the driver to fail is the disconnection.

So why not report that to the caller?

Reply:
I don't think I can improve my understanding without a full revisit of how throw works, what common practice are. I was not requesting a full stack trace just a "serial protocol logic" trace, that is not detailing the inner guts.